eox_nelp/notifications/tasks.py

```diff
@@ -0,0 +1,88 @@
+"""Notification tasks. Contains all tasks related with the notifications features.
+
+tasks:
+    create_course_notifications: Creates UpcomingCourseDueDate records based on due dates.
+"""
+import datetime
+
+from celery import shared_task
+from django.utils import timezone
+from opaque_keys.edx.keys import CourseKey
+
+from eox_nelp.edxapp_wrapper.course_overviews import CourseOverview
+from eox_nelp.edxapp_wrapper.modulestore import modulestore
+from eox_nelp.notifications.models import UpcomingCourseDueDate
+
+
+@shared_task
+def create_course_notifications(course_id):
+    """Creates UpcomingCourseDueDate records based on due dates. This considers the following rules:
+
+        1. The course must have the notification feature active, add the following dict to other_course_settings
+           to do that.
+
+            {
+                "notifications": {
+                    "active": true,
+                    "days_in_advance": [
+                        1,
+                        2,
+                        3
+                    ]
+                }
+            }
+
+        2. If days_in_advance is not set, the default value will be [7]
+        3. If the subsection doesn't have a due date, nothing will happen.
+        4. If the subsection already has UpcomingCourseDuedates records and they has not been sent, they will be
+           deleted.
+        5. If the subsection already has UpcomingCourseDuedates records and they has been sent, this will preserve
+           that data to keep a register.
+        6. If the notification date is in the past, the record won't be created.
+
+    Args:
+        course_id (str): Unique course identifier.
+    """
+    course_key = CourseKey.from_string(course_id)
+    course = modulestore().get_course(course_key)
+
+    notifications = course.other_course_settings.get("notifications", {})
+
+    if not notifications.get("active"):
+        return
+
+    now = timezone.now()
+    days_in_advance = notifications.get("days_in_advance")
+
+    if not days_in_advance:
+        days_in_advance = [7]
+
+    if not isinstance(days_in_advance, list):
+        days_in_advance = [days_in_advance]
+
+    sub_sections = modulestore().get_items(course_key, qualifiers={"category": "sequential"})
+
+    for sub_section in sub_sections:
+        due_date = sub_section.due
+
+        if not due_date:
+            continue
+
+        UpcomingCourseDueDate.objects.filter(  # pylint: disable=no-member
```

Comment:
Why you have the others UpcomingCourseDueDate, that wasn't sent. To clean if there is a change?

Reply:
I mean each time the signal course_published is triggered it would delete the records, but It has to be that way if the  subsection didn't change?. IMay be is possible to compare the new due date of the block with the due_date of the record to check if deserves to delete the previous record. Because each time I create units or activate course published it's probably that the same record(subsection) of due_date, location_id, notification_date is deleted and recreated..

Reply:
please check last commit